SPIFlashLogger.class.nut

```diff
@@ -76,53 +114,185 @@ class SPIFlashLogger {
-            if (_at_pos >= SPIFLASHLOGGER_SECTOR_SIZE) {
-                _at_sec = (_at_sec + 1) % _sectors;
-                _at_pos = SPIFLASHLOGGER_SECTOR_META_SIZE;
+            objPos += secRemaining;
+            objRemaining -= secRemaining;
+            _atPos += secRemaining;
+            if (_atPos >= SPIFLASHLOGGER_SECTOR_SIZE) {
+                _atSec = (_atSec + 1) % _sectors;
+                _atPos = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
             }
-        } while (obj_remaining > 0);
+        } while (objRemaining > 0);
 
         _disable();
     }
 
+    // Reads objects from the logger synchronously,
+    // returning a single log object for the specified *index*.
+    //
+    // Returns:
+    // Deserialized object or null
+    //    - the most recent object when `index == -1`,
+    //    - the oldest object when `index == 1`,
+    //    - *null* when the value of *index* is greater than the number of logs,
+    //    - throws an exception when `index == 0`.
+    //
+    function readSync(index = 0) {
+        // Unexpected index value
+        if (index == 0)
+            throw "Invalid argument";
+
+        // identify first sector to read
+        local seekTo = math.abs(index);
+        local count = 0;
+        local i = 0;
+
+        // _atSec - indicates the current write position, therefore
+        // for the index > 0 it is necessary to step forward (skip current sector)
+        // to find first not empty sector
+        // For the index < 0, it is possible to start couting from the current sector
+        while (i < _sectors) {
+            // convert sector index `i`, ordered by recency, to physical `sector`, ordered by position on disk
+            local sector;
+            if (index > 0) {
+                sector = (_atSec + i + 1) % _sectors;
+            } else {
+                sector = (_atSec - i + _sectors) % _sectors;
+            }
+
+            ++i;
+
+            // Read all objects start codes for current sector
+            // and write them into the blob
+            local objectsStartCodes = _getObjectsStartCodesForSector(sector);
+
+            // If there is no start codes in the current sector
+            // then switch to the next sector
+            if (objectsStartCodes == null || objectsStartCodes.len() == 0)
+                continue;
+
+            // negative step, go backwards
+            if (index < 0)
+                objectsStartCodes.seek(-2, 'e');
+
+            // Check that the number of start codes is enough
+            // otherwise decrease the seekTo count on a number of objects
+            // in the current sector and switch to the next sector
+            if (seekTo > objectsStartCodes.len() / 2) {
+              seekTo -= objectsStartCodes.len() / 2;
+              continue;
+            }
+
+            // seek for an object start code
+            if (objectsStartCodes.seek((seekTo - 1) * 2 * (index > 0 ? 1 : -1), 'c') == -1 || objectsStartCodes.eos() == 1) {
```

Comment:
Won't objectsStartCodes.eos() return -1 in case we are reading the last object in the sector? I guess no.

Reply:
Everything is fine: https://electricimp.com/docs/squirrel/blob/eos/
```
Returns
Integer — 1 if the pointer has reached the end of the blob, otherwise null
```